File: array/158.py
def wordPattern(pattern: str, s: str) -> bool:
    mapping_1, mapping_2 = {}, {}
    words = s.strip().split()

    if len(pattern) != len(words):
        return False

    for idx, c in enumerate(pattern):
        condition_1 = c in mapping_1 and mapping_1[c] != words[idx]
        condition_2 = words[idx] in mapping_2 and mapping_2[words[idx]] != c
        if condition_1 or condition_2:
            return False
        else:
            mapping_1[c] = words[idx]
            mapping_2[words[idx]] = c

    return True


print(wordPattern("abba", "dog cat cat fish"))

File: array/158.cpp
//
// Created by fzamora on 23-07-24.
//

#include<bits/stdc++.h>

using namespace std;

vector<string> split(string &s) {
    stringstream formatter(s);
    string cad;
    vector<string> splittedString;

    while(formatter >> cad) {
        splittedString.emplace_back(cad);
    }

    return splittedString;
}

bool wordPattern(string pattern, string s) {
    unordered_map<char, string> um_1;
    unordered_map<string, char> um_2;
    vector<string> words = split(s);

    if(words.size() != pattern.size())
        return false;

    for(int i = 0; i < pattern.size(); ++i) {
        bool condition_1 = um_1.contains(pattern[i]) && um_1[pattern[i]] != words[i];
        bool condition_2 = um_2.contains(words[i]) && um_2[words[i]] != pattern[i];
        if(condition_1 || condition_2) {
            return false;
        } else {
            um_1[pattern[i]] = words[i];
            um_2[words[i]] = pattern[i];
        }
    }

    return true;
}

int main() {
    string s = "dog dog dog dog";
    string pattern = "abba";

    cout<<boolalpha<<wordPattern(pattern, s);

    return 0;
}
